keras_cv/ops/box_matcher.py

```diff
@@ -0,0 +1,235 @@
+# Copyright 2022 The KerasCV Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"Argmax-based box matching"
+
+from typing import List
+from typing import Tuple
+
+import tensorflow as tf
+
+
+class ArgmaxBoxMatcher:
+    """Box matching logic based on argmax of highest value (e.g., iou).
+
+    This class computes matches from a similarity matrix. Each row will be
+    matched to at least one column, the matched result can either be positive
+     / negative, or simply ignored depending on the setting.
+
+    The settings include `thresholds` and `match_values`, for example if:
+    1) thresholds=[negative_threshold, positive_threshold], and
+       match_values=[negative_value, ignore_value, positive_value]: the rows will
+       be assigned to positive_value if its argmax result is above
+       positive_threshold; the rows will be assigned to negative_value if its
+       argmax result is below negative_threshold, and the rows will be assigned
+       to ignore_value if its argmax result is between negative_threshold and
+       positive_threshold.
+    2) thresholds=[negative_threshold, positive_threshold], and
+       match_values=[ignore_value, negative_value, positive_value]: the rows will
+       be assigned to positive_value if its argmax result is above
+       positive_threshold; the rows will be assigned to ignore_value if its
+       argmax result is below negative_threshold, and the rows will be assigned
+       to negative_value if its argmax result is between negative_threshold and
+       positive_threshold.
+    3) thresholds=[positive_threshold], and
+       match_values=[negative_values, positive_value]: the rows will be assigned to
+       positive value if its argmax result is above positive_threshold; the rows
+       will be assigned to negative_value if its argmax result is below
+       negative_threshold.
+
+    Usage:
+
+    ```python
+    box_matcher = keras_cv.ops.ArgmaxBoxMatcher([0.3, 0.7], [-1, 0, 1])
+    iou_metric = keras_cv.bounding_box.compute_iou(anchors, gt_boxes)
+    matched_columns, matched_match_values = box_matcher(iou_metric)
+    cls_mask = tf.less_equal(matched_match_values, 0)
+    ```
+
+    TODO(tanzhenyu): document when to use which mode.
+
+    """
+
+    def __init__(
+        self,
+        thresholds: List[float],
+        match_values: List[int],
+        force_match_for_each_col: bool = False,
+    ):
+        """Constructs ArgmaxBoxMatcher.
```

Comment:
nit: Lets move the __init__ docstring to the class docstring to be consistent with Keras/Cv/NLP

Reply:
Done.